app/domain/services/fetch_course_events/service.rb

```diff
@@ -679,31 +679,37 @@ def fetch_and_process_course_events(courses, event_types, restart, current_time)
         assignment_uuids_by_exercise_calculation_uuid[exercise_calculation_uuid] << uuid
       end
 
-      algorithm_exercise_calculations = AlgorithmExerciseCalculation
-        .where(exercise_calculation_uuid: exercise_calculation_uuids)
-        .to_a
-
       used_exercise_calculations_uuids = []
-      algorithm_exercise_calculations.each do |algorithm_exercise_calculation|
-        exercise_calculation_uuid = algorithm_exercise_calculation.exercise_calculation_uuid
-        algorithm_exercise_calculation.pending_assignment_uuids = (
-          algorithm_exercise_calculation.pending_assignment_uuids +
-          assignment_uuids_by_exercise_calculation_uuid[exercise_calculation_uuid]
-        ).uniq
-
+      algorithm_exercise_calculation_values = []
+      AlgorithmExerciseCalculation
+        .where(exercise_calculation_uuid: exercise_calculation_uuids)
+        .pluck(:uuid, :exercise_calculation_uuid, :pending_assignment_uuids)
+        .each do |uuid, exercise_calculation_uuid, pending_assignment_uuids|
         used_exercise_calculations_uuids << exercise_calculation_uuid \
-          if used_algorithm_exercise_calculation_uuids.include? algorithm_exercise_calculation.uuid
+          if used_algorithm_exercise_calculation_uuids.include? uuid
+
+        assignment_uuids = assignment_uuids_by_exercise_calculation_uuid[exercise_calculation_uuid]
+        # Don't bother updating records where assignment_uuids is empty
+        algorithm_exercise_calculation_values << [
+          uuid, (pending_assignment_uuids + assignment_uuids).uniq
+        ] unless assignment_uuids.empty?
       end
 
       ExerciseCalculation.where(uuid: used_exercise_calculations_uuids)
                          .ordered_update_all(is_used_in_assignments: true) \
         unless used_exercise_calculations_uuids.empty?
 
-      results << AlgorithmExerciseCalculation.import(
-        algorithm_exercise_calculations, validate: false, on_duplicate_key_update: {
-          conflict_target: [ :uuid ], columns: [ :pending_assignment_uuids ]
-        }
-      )
+      unless algorithm_exercise_calculation_values.empty?
+        algorithm_exercise_calculation_uuids = algorithm_exercise_calculation_values.map(&:first)
+        AlgorithmExerciseCalculation.update_all(
+          <<~UPDATE_SQL
+            "pending_assignment_uuids" = "values"."pending_assignment_uuids"
+            FROM (#{ValuesTable.new(algorithm_exercise_calculation_values)}) AS "values"
+              ("uuid", "pending_assignment_uuids")
+            WHERE "algorithm_exercise_calculations"."uuid" = "values"."uuid"::uuid
+          UPDATE_SQL
+        )
+      end
     end
```

Comment:
Do we need to use a `FOR UPDATE` lock to query these?

Reply:
There's another lock in a related record already, but it might be a good idea to add one here.. I'll add it.